extracted/install_base-8.30/sbin/instck/instck.man
.\" $Id: instck.man,v 8.8 2010/08/13 19:10:52 ksb Exp $
.\" Copyright 1990 Purdue Research Foundation, West Lafayette, Indiana
.\" 47907.  All rights reserved.
.\" $Compile: Display%h
.\" $Display: ${groff-groff} -Tascii -tbl -man %f | ${PAGER:-less}
.\" $Install: %b -mDeinstall %o %f && cp %f $DESTDIR/usr/local/man/man8/instck.8
.\" $Deinstall: ${rm-rm} -f $DESTDIR/usr/local/man/[cm]a[nt]8/instck.8*
.\" $Laser: ${tbl-tbl} %f | ${ltroff-ltroff} -man
.\"
.\" Written by Kevin S Braunsdorf, ksb@cc.purdue.edu, purdue!ksb
.\"	       Jeff Smith, jsmith@cc.purdue.edu, purdue!jsmith
.\"
.\" This software is not subject to any license of the American Telephone
.\" and Telegraph Company or the Regents of the University of California.
.\"
.\" Permission is granted to anyone to use this software for any purpose on
.\" any computer system, and to alter it and redistribute it freely, subject
.\" to the following restrictions:
.\"
.\" 1. Neither the authors nor Purdue University are responsible for any
.\"    consequences of the use of this software.
.\"
.\" 2. The origin of this software must not be misrepresented, either by
.\"    explicit claim or by omission.  Credit to the authors and Purdue
.\"    University must appear in documentation and sources.
.\"
.\" 3. Altered versions must be plainly marked as such, and must not be
.\"    misrepresented as being the original software.
.\"
.\" 4. This notice may not be removed or altered.
.TH INSTCK 1L PUCC
.SH NAME
instck \- check, repair, or record the modes of files
.SH SYNOPSIS
\fBinstck\fP [\-\fBdlLqSv\fP] [\-\fBC\fP\~\fIchecklist\fP] [\-\fBg\fP\~\fIgroup\fP] [\-\fBm\fP\~\fImode\fP] [\-\fBo\fP\~\fIowner\fP] \fIdirectories\fP
.br
\fBinstck\fP \-\fBi\fP [\-\fBdlLSvy\fP] [\fB\-C\fP\~\fIchecklist\fP] [\fB\-g\fP\~\fIgroup\fP] [\fB\-m\fP\~\fImode\fP] [\fB\-o\fP\~\fIowner\fP] \fIdirectories\fP
.br
\fBinstck\fP \fB\-G\fP [\-\fBR\fP | \-\fBd\fP] [\fB\-lLSvx\fP] [\-\fBg\fP\~\fIgroup\fP] [\fB\-m\fP\~\fImode\fP] [\fB\-o\fP\~\fIowner\fP] \fIdirectories\fP
.br
\fBinstck\fP \fB\-h\fP
.br
\fBinstck\fP \fB\-V\fP

.SH DESCRIPTION
.PP
.I Instck
inspects the modes of files in a directory structure.
.PP
It has two major methods of operation:
.br
 \fIscanning\fP, to generate a "good" list, and
 \fIchecking\fP, to see if the "good" list is being honored.
.br
.PP
For instance, your system might have the file \fI/usr/local/lib/install.cf\fP,
which would contain a "good" list that states /bin/sh should be installed mode
0755, owned by user root and group wheel. In a weak moment, your local SA used
install to put a new version of sh on the machine via:
.br
	install \-m7555 \-o root \-g wheel sh /bin
.br
(note the extra \`5\').  Instck would find this error if it were to check
/bin/sh against your "good" list.
.PP
In checking mode, the default action is to report (not act on) files that do
not comply with the expected values from the checklist.  The default checklist
is /usr/local/lib/install.cf.
.SH OPTIONS
.TP
.BI \-C checklist
Search \fIchecklist\fP for the expressions to match the
installed files, see \fIinstall.cf\fP(5L).
.TP
.B \-d
Do not check the contents of a directory, as in \fIls\fP(1).
.TP
.BI \-g group
Set the default group for all installed files.
.TP
.B \-G
This option will generate a checklist file for the given directories.
\fBInstck\fP outputs some crass comments on the end of the lines describing
each node, so some editing is recommended.
However one should reflect carefully on these before deleting them.
.TP
.B \-h
Print a summary of \fIinstck\fP's usage.
.TP
.B \-i
Prompt the user with suggested shell commands, for example if /tmp was
the wrong mode \fIinstck\fP might prompt with:
.sp 1
	instck: `/tmp' mode 0777 doesn't have bits to match drwxrwxrwt (1777)
.br
	instck: chmod 1777 /tmp [nfhqy]
.sp 1
A reply of `y' or `Y' will run the proposed command, `f' will skip to
the next file.  For other choices see `h'.  If your job holds little value,
`-y' will make subsequent prompts assume a `y'.  Careful, this one bites.
.TP
.B \-l
A long list will be made of all the files that didn't match any rule.
Under \-\fBG\fP generate no default pattern, list all nodes.
.TP
.B \-L
Allows `link format' in the configuration file.  Link format records
where a hardlink points or where a symbolic link points.  This is quite
useful for programs that carry a link with them, like \fBvi\fB and \fBex\fP.
.TP
.BI \-m mode
Set the default mode for all installed files.
.TP
.BI \-o owner
Set the default owner for installed files.
.TP
.B \-R
Recursively descend from each directory given into all subdirectories,
like \fIls\fP(1).
.TP
.B \-S
Run as if \fIgeteuid\fP(2) returned zero.
.TP
.B \-q
Suppress warnings about dangerous unchecked files.
.TP
.B \-v
Be more verbose.  Under \-\fBG\fP do not use double quote (") to
copy values from the preceding lines.
.TP
.B \-V
Give version information about \fIinstck\fP itself.
.TP
.B \-x
In the \-\fBgG\fP output, separate plain files by execute bit.  Those
with any execute bits set will be higher in the output.  This tends to
allow more of the double quote short-hand.
.TP
.B \-y
Under \-\fBi\fP fake a `yes' response to each confirmation,
like \fIfsck\fP(8).
Note that some operations are too dangerous to allow this option to
force their action.  For these few this option changes the default
answer to yes.  The author thinks this might cross the fine
line between clever and stupid.
.SH EXAMPLES
.TP
\fBinstck\fP /bin /usr/bin /usr/ucb /usr/local/bin /etc
Report inconsistent installations in the listed directories, using
/usr/local/lib/install.cf as the reference point.
.TP
\fBinstck\fP \-v /etc
Report inconsistent installations in /etc, be very verbose.  Since the default
checklist doesn't have much specific information about files in /etc this may
not provide much useful information.
.TP
\fBinstck\fP \-GdLx /etc/*  > /tmp/etc.cf
Generate a list of only the directories in /etc/. Like \fIls\fP, -d and -R are
exclusive to one another.
.TP
\fBinstck\fP \-GRLx /usr > /tmp/usr.cf
Generate a recursive checklist for /usr and all the files in /usr.
.TP
\fBinstck\fP \-ilvC/tmp/usr.cf /usr
Use the previously generated checklist to check /usr for changes.  Prompt for
what repairs might be required.
.TP
\fBinstck\fP \-V
Report version information, e.g.
.br
.RS
.TS
l s s
l s s
l l l.
instck: version: $\&Id: instck.m,v 8.22 2006/12/30 03:21:06 ksb Exp $
instck: configuration file: /usr/local/etc/install.cf
instck: defaults: owner=root	group=binary	mode=\-rwxr\-xr\-x
.TE
.RE
.SH FILES
.TS
l l.
/usr/local/lib/instck.cf	the default \fIchecklist\fP file
.TE
.SH AUTHOR
KS Braunsdorf, NPC Guild.org
.br
console at-hate-spam ksb.npcguild.org
.br
Copyright \*(co 1990 Purdue Research Foundation.  All rights reserved.
This is a modified version of that code-base.
.SH SEE ALSO
ls(1), chgrp(1), chmod(1), install(1L), ranlib(1), strip(1),
geteuid(2),
syslog(3),
install.cf(5L), purge(8),
chown(8), fsck(8)


extracted/install_base-8.30/sbin/instck/instck.m
# mkcmd script for instck
# $Id: instck.m,v 8.28 2009/12/10 18:45:46 ksb Exp $
require "std_help.m"

from '<sys/param.h>'
from '<sys/types.h>'
from '<sys/stat.h>'
from '<sys/file.h>'
from '<sysexits.h>'
from '<syslog.h>'
from '<ar.h>'

from '"machine.h"'
from '"configure.h"'
from '"install.h"'
from '"main.h"'
from '"special.h"'
from '"syscalls.h"'
from '"instck.h"'

from '<errno.h>'
from '<pwd.h>'
from '<grp.h>'
from '"gen.h"'

vector "apcHelp"
getenv "INSTCK"
basename "instck" ""

key "instck_paths" 8 init {
	"chgrp"
	"chown"
	"strip"
	"chmod"
	"cmp"
	"rm"
	"rmdir"
	"ln"
	"mv"
	"sh"
}
key "install_path" 8 init {
	"%D/install/pv"		# /usr/local/bin/install, we hope
}

key "purge_path" 8 init {
	"%D/purge/pv"		# /usr/local/etc/purge, or from /usr/local/sbin
}

# if we can't find ranlib run /bin/true on the file...
key "instck_ranlib" 8 init {
	"ranlib" "true"
}
# if we can't find chflags fail with /bin/false
key "instck_chflags" 8 init {
	"chflags" "false"
}
require "instck.mh"

before {
	named "Before"
	update "%n();"
}

# general switches
char* 'C' {
	named "pcSpecial"
	init "CONFIG"
	param "config"
	help "use config to check for special files"
}
boolean 'd' {
	named "fDirs"
	init "1"
	help "do not scan given directories for files"
}
boolean 'i' {
	boolean 'y' {
		named "fYesYes"
		help "fake a `yes' response to each interactive query"
	}
	named "fInteract"
	help "interactively repair the installed files"
}
boolean 'l' {
	named "fLongList"
	help "list left over files (those that were not checked)"
}
boolean 'L' {
	named "fLinks"
	help "record hard links in link format"
}
boolean 'v'{
	named "fVerbose"
	help "be more verbose"
}
boolean 'S' {
	named "bHaveRoot"
	help "run as if we were the superuser"
}
boolean 'q' {
	named "fQuiet"
	init "0"
	help "supress warnings about dangerous unchecked files"
}
# install uses this one, but we do not, so emulate the flag
#boolean 'n' {
#	named "fTrace"
#	help "do not really take any destructive actions"
#}
integer variable "fTrace" {
	init "FALSE"
}

# modes
char* 'm' {
	named "pcMode"
	init "DEFMODE"
	param "mode"
	help "set the default mode for installed files"
}
char* 'o' {
	named "pcGenOwner"
	param "owner"
	help "set the default owner for installed files"
}
char* 'g' {
	named "pcGenGroup"
	param "group"
	help "set the default group for all files"
}

# major modes
boolean 'G' {
	named "fGen"
	exclude "iC"
	boolean 'x' {
		named "fByExec"
		help "separate plain files by execute bit"
	}
	boolean 'R' {
		exclude "d"
		named "fRecurse"
		help "descend into all directories"
	}
	help "generate a check list for the given dirs on stdout"
}
boolean 'V' {
	named "fVersion"
	aborts "break;"
	help "show version information"
}

after {
	named "After"
	update "%n();"
}

list {
	named "DoIt"
	param "files"
	help "check these or generate list for them"
}

exit {
	update "/* cleanup and exit */"
	aborts "Exit();exit(0);"
}

# we are su'd to root
char* variable "pcGuilty" {
}
char* variable "pcDefMode" {
}
integer named "iExitCode" {
	init "EX_OK"
	help "the exit code if nothing fatal happens"
}

%i
#if USE_ELF
#if USE_ELFABI
#include <elf_abi.h>
#else
#include <libelf.h>
#endif
#else
#if USE_SYS_EXECHDR_H
#include <sys/exechdr.h>
#else
#include <a.out.h>
#endif
#endif

#if HAVE_NDIR
#include <ndir.h>
#else
#if HAVE_DIRENT
#if USE_SYS_DIRENT_H
#include <sys/dirent.h>
#else
#include <dirent.h>
#endif
#else
#include <sys/dir.h>
#endif
#endif

/* old.h needs DIR defined before it is includeed
 */
#include "old.h"

char copyright[] =
   "@(#) Copyright 1990 Purdue Research Foundation.\nAll rights reserved.\n\
Modified version by ksb";
%%

%%

static char acMBuf[24];		/* mode buffer				*/

/* setup instck
 */
static void
Before()
{
	/* are we the superuser
	 */
	bHaveRoot = 0 == geteuid();
#if USE_ELF
#if USE_ELFABI
	/* cannot check the library itself */
#else
	if (EV_NONE == elf_version(EV_CURRENT)) {
		fprintf(stderr, "%s: wrong ELF version (%d)\n", progname, EV_CURRENT);
		exit(EX_PROTOCOL);
	}
#endif
#endif
}

/* fix any silly option mismatches					(ksb)
 * ``Be liberal with what you accept, strict in what you output.''
 */
static void
After()
{
	register char *pcDot;	/* joe.group style	*/
	auto int iM, iO;	/* mode buffers		*/

	/* set the files we need, instck the dirs we were asked to, end clean
	 */
	(void)setpwent();
	(void)setgrent();

	if ((char *)0 == pcGenGroup || '\000' == pcGenGroup[0]) {
		pcGenGroup = (char *)0;
	}
	if ((char *)0 == pcGenOwner || '\000' == pcGenOwner[0]) {
		pcGenOwner = (char *)0;
	} else if ((char *)0 != (pcDot = strchr(pcGenOwner, '.'))) {
		if ((char *)0 != pcGenGroup) {
			fprintf(stderr, "%s: group given with both -g and -o?\n", progname);
			exit(EX_USAGE);
		}
		*pcDot++ = '\000';
		if ('\000' != *pcDot) {
			pcGenGroup = pcDot;
		}
	}
	if (bHaveRoot) {
		if ((char *)0 == pcGenGroup) {
			pcGenGroup = DEFGROUP;
		}
		if ((char *)0 == pcGenOwner) {
			pcGenOwner = DEFOWNER;
		}
	}

	InitCfg(pcGenOwner, pcGenGroup);

	pcDefMode = acMBuf;
	acMBuf[0] = '-';
	if ((char *)0 == pcMode || '\000' == pcMode[0]) {
		pcMode = (char *)0;
	} else {
		CvtMode(pcMode, & iM, & iO);
		ModetoStr(pcDefMode+1, iM, iO);
	}
}


/* passed the list of files to handle					(ksb)
 */
static void
DoIt(argc, argv)
int argc;
char **argv;
{
	extern char *getenv();
	argc = ElimDups(argc, argv);

	if (fVersion) {
		static char acInh[] = "inherited";

		printf("%s: version: $Id: instck.m,v 8.28 2009/12/10 18:45:46 ksb Exp $\n", progname);
		printf("%s: configuration file: %s\n", progname, pcSpecial);
#if defined(INST_FACILITY)
		printf("%s: syslog facility: %d\n", progname, INST_FACILITY);
#endif
		printf("%s: defaults: owner=%-10s group=%-10s mode=%s\n", progname, (struct passwd *)0 != pwdDef ? pwdDef->pw_name : acInh, (struct group *)0 != grpDef ? grpDef->gr_name : acInh, (char *)0 != pcMode ? pcDefMode : acInh);
		return;
	}

	if (fGen) {
		GenCk(argc, argv);
		return;
	}

#if defined(INST_FACILITY)
	if (bHaveRoot) {
		extern char *getlogin();
		(void)openlog(progname, 0, INST_FACILITY);
		pcGuilty = getlogin();
		if ((char *)0 == pcGuilty || '\000' == *pcGuilty) {
			pcGuilty = getenv("USER");
			if ((char *)0 == pcGuilty || '\000' == *pcGuilty)
				pcGuilty = getenv("LOGNAME");
			if ((char *)0 == pcGuilty || '\000' == *pcGuilty)
				pcGuilty = "root\?";
		}
	}
#endif

	argc = FilterOld(argc, argv, & CLCheck);
	InstCk(argc, argv, pcSpecial, & CLCheck);

#if defined(INST_FACILITY)
	if (bHaveRoot && fInteract) {
		(void)closelog();
	}
#endif
}

/* checkout for now
 */
static void
Exit()
{
	(void)endpwent();
	(void)endgrent();
	exit(iExitCode);
}
